libs/esp32boot/esp32init.erl

```diff
@@ -0,0 +1,394 @@
+        {failed_app_start, undefined} ->
+            start_dev_mode()
+    end.
+
+start_dev_mode() ->
+    avm_pubsub:start(default_pubsub),
+    spawn(?MODULE, maybe_start_network, []),
+    started.
+
+%%
+%% Boot handling
+%%
+
+% TODO: add support for multiple apps
+% /dev/partition/by-name/app1.avm
+% /dev/partition/by-name/app2.avm
+
+boot() ->
+    BootPath = get_boot_path(),
+    case atomvm:add_avm_pack_file(BootPath, [{name, app}]) of
+        ok ->
+            StartModule = get_start_module(),
+            maybe_start_dev_mode(ok),
+            StartModule:start();
+        {error, Reason} ->
+            io:format("Failed app start: ~p.~n", [Reason]),
+            maybe_start_dev_mode(failed_app_start)
+    end.
+
+get_boot_path() ->
+    case esp:nvs_get_binary(atomvm, boot_path) of
+        undefined ->
+            "/dev/partition/by-name/main.avm";
+        Path ->
+            Path
+    end.
+
+get_start_module() ->
+    case esp:nvs_get_binary(atomvm, start_module) of
+        undefined ->
+            case atomvm:get_start_beam(app) of
+                error ->
+                    main;
+                {ok, ModuleNameWithExt} when is_binary(ModuleNameWithExt) ->
+                    Len = byte_size(ModuleNameWithExt) - byte_size(<<".beam">>),
+                    ModuleName = binary:part(ModuleNameWithExt, 0, Len),
+                    erlang:binary_to_atom(ModuleName, latin1)
+            end;
+        Module ->
+            erlang:binary_to_atom(Module, latin1)
+    end.
+
+%%
+%% Network management
+%%
+
+erase_net_config() ->
+    io:format("Erasing net config.~n"),
+    esp:nvs_erase_key(atomvm, sta_ssid),
+    esp:nvs_erase_key(atomvm, sta_psk).
+
+save_net_config(SSID, Pass) ->
+    io:format("Saving config: SSID: ~p Pass: ~p.~n", [SSID, Pass]),
+    esp:nvs_set_binary(atomvm, sta_ssid, erlang:list_to_binary(SSID)),
+    esp:nvs_set_binary(atomvm, sta_psk, erlang:list_to_binary(Pass)).
+
+get_net_config() ->
+    case esp:nvs_get_binary(atomvm, sta_ssid) of
+        undefined ->
+            get_default_net_config();
+        SSID ->
+            case esp:nvs_get_binary(atomvm, sta_psk) of
+                undefined ->
+                    get_default_net_config();
+                Psk ->
+                    get_net_config(SSID, Psk)
+            end
+    end.
+
+get_default_net_config() ->
+    Creds = [
+        {ssid, ?DEFAULT_AP_SSID},
+        {psk, ?DEFAULT_AP_PSK}
+    ],
+    {wait_for_ap, Creds}.
+
+get_net_config(SSID, Psk) ->
+    Creds = [
+        {ssid, SSID},
+        {psk, Psk}
+    ],
+    {wait_for_sta, Creds}.
+
+maybe_start_network() ->
+    case esp:nvs_get_binary(atomvm, wlan_enabled) of
+        undefined ->
+            start_network();
+        <<"always">> ->
+            start_network();
+        <<"never">> ->
+            not_started
+    end.
+
+start_network() ->
+    io:format("Starting network...~n"),
+    {WaitFunc, Creds} = get_net_config(),
+    case network:WaitFunc(Creds) of
```

Comment:
This isn't really the right function to use, if you are trying to build a robust application.  For example, if the Wifi server restarts, the TCP connects may need to be restarted, etc.  But if this is just for "dev mode" and not a serious mobile application, it's probably okay.

Reply:
Indeed, it is sub-optimal, let's iterate over this with more additional PRs.